storm-core/src/jvm/org/apache/storm/scheduler/EvenScheduler.java

```diff
@@ -0,0 +1,168 @@
+
+    public static List<WorkerSlot> sortSlots(List<WorkerSlot> availableSlots) {
+        //For example, we have a three nodes(supervisor1, supervisor2, supervisor3) cluster:
+        //slots before sort:
+        //supervisor1:6700, supervisor1:6701,
+        //supervisor2:6700, supervisor2:6701, supervisor2:6702,
+        //supervisor3:6700, supervisor3:6703, supervisor3:6702, supervisor3:6701
+        //slots after sort:
+        //supervisor3:6700, supervisor2:6700, supervisor1:6700,
+        //supervisor3:6701, supervisor2:6701, supervisor1:6701,
+        //supervisor3:6702, supervisor2:6702,
+        //supervisor3:6703
+
+        if (availableSlots != null && availableSlots.size() > 0) {
+            // group by node
+            Map<String, List<WorkerSlot>> slotGroups = new TreeMap<String, List<WorkerSlot>>();
+            for (WorkerSlot slot : availableSlots) {
+                String node = slot.getNodeId();
+                List<WorkerSlot> slots = null;
+                if(slotGroups.containsKey(node)){
+                   slots = slotGroups.get(node);
+                }else{
+                   slots = new ArrayList<WorkerSlot>();
+                   slotGroups.put(node, slots);
+                }
+                slots.add(slot);
+            }
+
+            // sort by port: from small to large
+            for (List<WorkerSlot> slots : slotGroups.values()) {
+                Collections.sort(slots, new Comparator<WorkerSlot>() {
+                    @Override
+                    public int compare(WorkerSlot o1, WorkerSlot o2) {
+                        return o1.getPort() - o2.getPort();
+                    }
+                });
+            }
+
+            // sort by available slots size: from large to small
+            List<List<WorkerSlot>> list = new ArrayList<List<WorkerSlot>>(slotGroups.values());
+            Collections.sort(list, new Comparator<List<WorkerSlot>>() {
+                @Override
+                public int compare(List<WorkerSlot> o1, List<WorkerSlot> o2) {
+                    return o2.size() - o1.size();
+                }
+            });
+
+            return Utils.interleaveAll(list);
+        }
+
+        return null;
+    }
+
+    public static Map<WorkerSlot, List<ExecutorDetails>> getAliveAssignedWorkerSlotExecutors(Cluster cluster, String topologyId) {
+        SchedulerAssignment existingAssignment = cluster.getAssignmentById(topologyId);
+        Map<ExecutorDetails, WorkerSlot> executorToSlot = null;
+        if (existingAssignment != null) {
+            executorToSlot = existingAssignment.getExecutorToSlot();
+        }
+
+        return Utils.reverseMap(executorToSlot);
+    }
+
+    private static Map<ExecutorDetails, WorkerSlot> scheduleTopology(TopologyDetails topology, Cluster cluster) {
+        List<WorkerSlot> availableSlots = cluster.getAvailableSlots();
+        Set<ExecutorDetails> allExecutors = (Set<ExecutorDetails>) topology.getExecutors();
```

Comment:
Here too it would be nice to just create a new Set from the Collection returned by topology.getExecutors.